service-base/src/main/java/org/eclipse/hono/service/registration/CompleteBaseRegistrationService.java

```diff
@@ -0,0 +1,201 @@
+            final Future<RegistrationResult> result = Future.future();
+            getDevice(tenantId, deviceId, result.completer());
+            return result.map(res -> {
+                return request.getResponse(res.getStatus())
+                        .setDeviceId(deviceId)
+                        .setJsonPayload(res.getPayload())
+                        .setCacheDirective(res.getCacheDirective());
+            });
+        }
+    }
+
+    private Future<EventBusMessage> processUpdateRequest(final EventBusMessage request) {
+
+        final String tenantId = request.getTenant();
+        final String deviceId = request.getDeviceId();
+        final JsonObject payload = getRequestPayload(request.getJsonPayload());
+
+        if (tenantId == null || deviceId == null) {
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST));
+        } else {
+            log.debug("updating registration information for device [{}] of tenant [{}]", deviceId, tenantId);
+            final Future<RegistrationResult> result = Future.future();
+            updateDevice(tenantId, deviceId, payload, result.completer());
+            return result.map(res -> {
+                return request.getResponse(res.getStatus())
+                        .setDeviceId(deviceId)
+                        .setCacheDirective(res.getCacheDirective());
+            });
+        }
+    }
+
+    private Future<EventBusMessage> processDeregisterRequest(final EventBusMessage request) {
+
+        final String tenantId = request.getTenant();
+        final String deviceId = request.getDeviceId();
+
+        if (tenantId == null || deviceId == null) {
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST));
+        } else {
+            log.debug("deregistering device [{}] of tenant [{}]", deviceId, tenantId);
+            final Future<RegistrationResult> result = Future.future();
+            removeDevice(tenantId, deviceId, result.completer());
+            return result.map(res -> {
+                return request.getResponse(res.getStatus())
+                        .setDeviceId(deviceId)
+                        .setCacheDirective(res.getCacheDirective());
+            });
+        }
+    }
+
+    /**
+     * Processes a request for a non-standard operation.
+     * <p>
+     * Subclasses should override this method in order to support additional, custom
+     * operations that are not defined by Hono's Device Registration API.
+     * <p>
+     * This default implementation simply returns a future that is failed with a
+     * {@link ClientErrorException} with an error code <em>400 Bad Request</em>.
+     *
+     * @param request The request to process.
+     * @return A future indicating the outcome of the service invocation.
+     */
+    protected Future<EventBusMessage> processCustomRegistrationMessage(final EventBusMessage request) {
```

Comment:
since this is the same as in `BaseRegistrationService` I don't think that we need to override it, do we?